File: Python/Binary Tree Maximum Path Sum.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
	maxAcrossRoot = -65536
	def maxPathSum(self, root):
		maxEndByRoot = self.getMax(root,self.maxAcrossRoot)
		return max(self.maxAcrossRoot,maxEndByRoot)
	def getMax(self,node,maxAcrossRoot):
		if node == None:
			return 0
		left = self.getMax(node.left,self.maxAcrossRoot)
		right = self.getMax(node.right,self.maxAcrossRoot)
		cMax = node.val
		if left > 0:
			cMax = cMax + left
		if right > 0:
			cMax = cMax + right
		self.maxAcrossRoot = max(self.maxAcrossRoot,cMax)
		return max(node.val,max(node.val+left,node.val+right))
		
		
#another solution# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
	def maxPathSum(self, root):
	

File: Java/Binary Tree Maximum Path Sum.java
class Binary Tree Maximum Path Sum{
	private int maxV = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        int nonCross = findPathSum(root);
        return Math.max(nonCross,this.maxV);
    }
    private int findPathSum(TreeNode root){
    	if(root == null)
    		return 0;
    	int left = findPathSum(root.left);
    	int right = findPathSum(root.right);
    	int cross = left + right + root.val;
    	int non_cross = Math.max(root.val,root.val+Math.max(left,right));
    	maxV = Math.max(cross,non_cross);
    	return non_cross;
    }
}